### ISSUE TITLE
Support for variable replacement

### ISSUE THREAD

**Issue by struberg** [eclipse/ConfigJSR#23]
In Apache DeltaSpike we support configure some.key=http://${bla.server}/some/Endpoint
where bla.server is yet another config entry.

https://issues.apache.org/jira/browse/DELTASPIKE-1126

This is also covered in the original ConfigJSR proposal:
https://github.com/struberg/javaConfig/blob/configValue/api/src/main/java/javx/config/ConfigValue.java#L54

Do we want to support this kind of functionality in JSR-382?

**Comment by kazumura** [comment]
I think yes.

If there a lot of derived keys, and
when you change the base key, you need to change all of derived keys manually.

>  some.base=foo
>  some.key1=foo/111
>  some.key2=foo/222
>   ...

Although your editor may have the word replacing feature,
you need to check them.

**Comment by struberg** [comment]
Hi Kenji, txs for the feedback.

From my experience with DeltaSpike in big banking projects it was extremely helpful.
We were able to keep the default config but just tweak the base url for the whole cluster with one entry. All that by just overriding a single entry via our self provided DatabaseConfigSource.
I imagine that such a cluster wide ConfigSource could even be part of an AppServer which targets Clustering as first class citizen.

**Comment by TFaga** [comment]
Agreed on this. We should definitely include some sort of string interpolation into config values while still keeping it simple.

In KumuluzEE we have the exact same system in place for variable replacements and its really useful especially since there the entire framework is configured through the config system, so reusing those values becomes really useful.

Support for this in the JSR would be great.

**Comment by jmesnil** [comment]
It might be worth considering also default values in the expression:

```
some.key=http://${blah.server:example.com}/some/endpoint
```

If a config property with the `blah.server` is found, it is used, otherwise the default value `example.com` is used.

**Comment by struberg** [comment]
To be honest, I don't like default values. Neither in this way nor in code.
With JavaConfig default values can easily be provided in a default property file (low priority). That way people also know that such a key even exists.

**Comment by Emily-Jiang** [comment]
I agree with @struberg ! Let's keep it simply in our first version.

**Comment by jmesnil** [comment]
ok, let's skip default for expressions.
However I think we should ensure that any config that is in an expression has an actual value.
For example, if we have:

```
some.key=http://${bla.server}/some/Endpoint
```

but there is no `bla.server` property configured, it should not result in `some.key` being evaluated to `http:///some/Endpoint`

We have 2 options:
1. raise an exception when we evaluate an expressions that has unconfigured property
2. consider the expression unconfigured (as its value can not be properly resolved)

wdyt?

**Comment by kazumura** [comment]
Option 1. is easier to debug.

**Comment by lilian-benoit** [comment]
Option 1 for Config#getValue
Option 2 for Config#getOptionalValue

**Comment by struberg** [comment]
This is already covered via the new ConfigAccessor feature resolved with #9
